### PR TITLE
Prevent expansion of cat indexing to avoid int64 intermediate

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127815

Fix for https://github.com/pytorch/pytorch/issues/127652

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/optimize_indexing.py [hunk 1]
```diff
@@ -116,3 +116,17 @@ def indexing_dtype_strength_reduction(loop_body: LoopBody):
...
+
+def all_indexing_expressable_in_32_bits(loop_body: LoopBody) -> bool:
+    """
+    Are all get_index and index_expr nodes values expressable in 32 bits
+    """
+    bv = loop_body.bounds().get_bounds()
+    for node in loop_body.get_nodes():
+        if node.target in ("get_index", "index_expr"):
+            if vr := bv.get(node, None):
+                if not range_expressable_in_32_bits(vr):
```
Comment (peterbell10) [inline torch/_inductor/optimize_indexing.py:129]: The current strategy is to use the hint values to determine if we should generate a 32-bit indexing kernel, and generate guards appropriately. This means we get optimized kernels even if the range analysis can't prove an upper bound, which happens with basically any dynamic shape.

I'm not sure relying on range analysis is going to work at all tbh.

File: torch/_inductor/codegen/simd.py [hunk 2]
```diff
@@ -1206,7 +1216,9 @@ def _get_buffer(name: str) -> Union[ir.Buffer, ir.TensorBox]:
         # conservative here.
         total_numel = numel * reduction_numel
 
-        if SIMDScheduling.can_use_32bit_indexing(total_numel, buffers):
+        if can_expr_in_32 and SIMDScheduling.can_use_32bit_indexing(
+            total_numel, buffers
+        ):
```
Comment (peterbell10) [inline torch/_inductor/codegen/simd.py:1221]: A bit part of the original issue is the use of `sympy.expand` to flatten indexing expressions. e.g. in the pointwise cat we have expressions like 
```
stride0 * (x0 - c0)
```
where this expression fits inside the input buffer and can use 32-bit indexing, but when we expand to
```
-c0 * stride0 + x0 * stride0
```
The intermediates of this expression can fall outside the valid range when the mask is true.

So, I wonder if a better fix would be to block the pointwise cat indexing from being expanded. e.g. you could define an `Identity` sympy function which won't get expanded.

File: torch/_inductor/optimize_indexing.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/optimize_indexing.py:129]: Yea, didn't forgot range analysis bails on dynamic shapes.

**Comment by shunting314** [comment]
Can you show the generated triton kernel before and after the fix so it's easier to understand the fix.

File: test/inductor/test_cuda_repro.py [hunk 3]
```diff
@@ -1204,6 +1204,46 @@ def forward(self, x):
             out2 = m(input_tensor)
             self.assertEqual(out, out2, atol=1e-3, rtol=1e-3)
 
+    def test_int64_index_intermediate(self):
+        def foo(split_1):
+            view_23 = torch.ops.aten.view.default(inp, [-1, 8192, 8192])
+            split_1 = torch.ops.aten.split.Tensor(view_23, 1024, 1)
```
Comment (shunting314) [inline test/inductor/test_cuda_repro.py:1210]: Clean up this a bit?

`split_1` argument is not used and being overwritten directly.

**Comment by eellison** [comment]
@shunting314  

previous codegen:

`tmp40 = tl.load(in_ptr0 + ((-2642411520) + x0 + (67108864*x1)), tmp39, other=0.0).to(tl.float32)`

new codegen:

`tmp40 = tl.load(in_ptr0 + (41943040 + x0 + (67108864*(x1 + ((-5)*(ks0 // 8192))))), tmp39, other=0.0).to(tl.float32)`

**Comment by shunting314** [comment]
So the new codegen works because we change the index computation order so we don't generate intermediate like '-2642411520'?

File: torch/_inductor/lowering.py [hunk 4]
```diff
@@ -1015,7 +1015,9 @@ def inner_fn(idx):
 
             # if we're concatting [4], [2]
             # when we index the second tensor for 5 we want to index 5 - 4
-            idx_load[dim] -= inputs_ranges[i][0]
+            # Use Identity to prevent expansion of index * stride to keep expression
```
Comment (shunting314) [inline torch/_inductor/lowering.py:1018]: So the root cause of the issue is

`-inputs_range[i][0] * stride` get underflow?

**Comment by shunting314** [comment]
I don't expect metric change on the dashboard since the PR only affect `pointwise_cat`. But it may worth check if the use of Identity cause the index expression being less efficient and results in the perf degradation for `pointwise_cat`

**Review by shunting314 — APPROVED** [review]

**Review by peterbell10 — APPROVED** [review]

**Pull request closed without merging**